### DISCUSSION TITLE
setValue in useEffect during tests causes "An update not wrapped in act"

### DISCUSSION THREAD

**Discussion by httpete** [react-hook-form/react-hook-form#9722] [Q&A]
I have a useEffect in my Form component like this:

```
  useEffect(() => {
    setValue(attributes.name, attributes.value, {
      shouldValidate: true,
    });
  }, [setValue, attributes]);
  
  ```
  And the shouldValidate: true piece is what trips up this Warning only in Jest tests.  The goal is if I receive a new value, to set it on the form input and validate the input. I have tried every combination of waitFor and screen.findBy as the doc says, but nothing makes the error stop as long as I do shouldValidate: true.

**Comment by leapful** [comment]
@httpete 

Please provide a codesandbox url or repo that can re-produce the issue. No one knows `what trips up this Warning only in Jest tests` is what kind of error since your code example only had a `useEffect` hook without any context, there is no way to help you on solving this problem.

Follow the template for asking question here: https://github.com/react-hook-form/react-hook-form/issues/2848

**Comment by httpete** [comment]
I have tried to be more careful with only setting the value and validating if it has changed, and this does work:

```
  useEffect(() => {
    const { name, value } = attributes;

    // If this is the initial mount, I don't want to set the value, as I do it in the defaultValue below.
    if (firstUpdate.current) {
      firstUpdate.current = false;

      return;
    }
    // If the currentValue of the field is the same that is on the form, I also don't want to update the value or validate.
    const currentValue = getValues(name);
    if (currentValue === value) {
      return;
    }
    // Now the server has pushed us a new value, I want to set it, and validate the input.
    setValue(name, value, {
      shouldValidate: true,
    });
  }, [setValue, getValues, attributes]);

**Reply by httpete** [reply]
The calling of setValue with shouldValidate in a useEffect trips the "act" warning. In a way, this showed me that it was being called too many times, but I can see how this could be a problem in certain use cases.
